### PR TITLE
[enhancement] transition listener registration from member string name to member string pointer

### PR DESCRIPTION
**Pull request by odysseus654**
Qt permits signals to be connected to slots using either a method pointer or the string name of the method.  This page describes the strengths of both: https://doc.qt.io/qt-5/signalsandslots-syntaxes.html

Registering interest in specific packet types is done through registering a listener with PacketReceiver.  Up to this point all registrations have been made by string, which required digging into the Qt reflection layer to confirm the method was compatible and indirectly invoke the method.

This patch changes all listener registrations to use functors referencing member pointers.  This enforces the existence and parameters of the method at compile-time rather than run-time.  There may be a small performance benefit from the reduced processing required, but the focus here is in getting away from using strings to identify members.

### CODE REVIEW

File: libraries/networking/src/ResourceCache.cpp [hunk 1]
```diff
@@ -16,8 +16,9 @@
 #include <cmath>
 #include <assert.h>
 
-#include <QThread>
-#include <QTimer>
+#include <QtCore/QMetaMethod>
```
Comment (HifiExperiments) [inline libraries/networking/src/ResourceCache.cpp:19]: same here

File: libraries/networking/src/PacketReceiver.h [hunk 2]
```diff
@@ -93,4 +135,42 @@ class PacketReceiver : public QObject {
     friend class OctreePacketProcessor;
 };
 
+template <class T>
+PacketReceiver::ListenerReferencePointer PacketReceiver::makeUnsourcedListenerReference(T* target, void (T::* slot)(QSharedPointer<ReceivedMessage>)) {
+    return QSharedPointer<UnsourcedListenerReference<T>>::create(target, slot);
```
Comment (HifiExperiments) [inline libraries/networking/src/PacketReceiver.h:140]: when I first saw all these shared pointers, I was a little nervous, but under the hood, this code is still doing exactly what the old code was doing, right?  the old code was just doing it in a silly way with strings.  the number of shared pointers + basically all the logic is pretty much staying the same?

File: domain-server/src/DomainGatekeeper.cpp [hunk 3]
```diff
@@ -17,7 +17,8 @@
 #include <openssl/x509.h>
 #include <random>
 
-#include <QDataStream>
+#include <QtCore/QDataStream>
+#include <QtCore/QMetaMethod>
```
[the remark this answers is not in the document]
Comment (odysseus654) [reply domain-server/src/DomainGatekeeper.cpp:21]: QMetaMethod as an include was removed from PacketReceiver.h.  That broke some files that included PacketReceiver.h, most likely making references to QMetaEnum, so #includes were added "downstream".

File: interface/src/ui/DomainConnectionModel.cpp [hunk 4]
```diff
@@ -9,7 +9,8 @@
 //
 
 #include "DomainConnectionModel.h"
-#include <QLoggingCategory>
+#include <QtCore/QLoggingCategory>
+#include <QtCore/QMetaMethod>
```
[the remark this answers is not in the document]
Comment (odysseus654) [reply interface/src/ui/DomainConnectionModel.cpp:13]: QMetaMethod as an include was removed from PacketReceiver.h.  That broke some files that included PacketReceiver.h, most likely making references to QMetaEnum, so #includes were added "downstream".

File: libraries/networking/src/ResourceCache.cpp (hunk 1 above)
Comment (odysseus654) [reply libraries/networking/src/ResourceCache.cpp:19]: QMetaMethod as an include was removed from PacketReceiver.h.  That broke some files that included PacketReceiver.h, most likely making references to QMetaEnum, so #includes were added "downstream".

File: domain-server/src/DomainServerSettingsManager.h [hunk 5]
```diff
@@ -211,6 +211,8 @@ private slots:
 
     /// guard read/write access from multiple threads to settings 
     QReadWriteLock _settingsLock { QReadWriteLock::Recursive };
+
+    friend class DomainServer;
```
[the remark this answers is not in the document]
Comment (odysseus654) [reply domain-server/src/DomainServerSettingsManager.h:215]: DomainServer is registering three private slots on this class.  When the slots were being identified by string through Qt reflection this wasn't a problem, but now that we're making direct member references it needs to have access.  So I figure either declare it as a friend or make the slots no longer private.  Suggestions are welcome, I don't really like either approach.

File: libraries/networking/src/PacketReceiver.h (hunk 2 above)
Comment (odysseus654) [reply libraries/networking/src/PacketReceiver.h:140]: Not quite sure I understand here, I am likely creating one new shared pointer representing the functor that is being used to call the listener, templated in order to properly use member function references (I believe Qt gets around this with something inside Q_OBJECT), and with two subclassed alternatives to handle the two types of function call parameters.  I attempted to approximate the old code as much as I obviously could, the differences should be about the difference between the two QObject::connect() syntaxes.

Comment (odysseus654) [reply libraries/networking/src/PacketReceiver.h:140]: As I am using subclassed functors they cannot be passed by value (tried that in an earlier version of this code) and pass-by-reference starts running into lifetime issues, thus using shared pointers.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-8fa9667_34e3d9dd.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-8fa9667.tar_34e3d9dd.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-8fa9667_34e3d9dd.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-8fa9667.tar_34e3d9dd.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-8fa9667_34e3d9dd.exe

**Comment by JulianGro** [comment]
How would we test this? Just see if it builds and runs?

**Comment by odysseus654** [comment]
At this point... probably.  Just make sure this code is at least as good as the code currently in the system.

**Comment by kasenvr** [comment]
This appears to touch quite a few parts of the system, especially the server. So that would have to be tested.

**Comment by odysseus654** [comment]
It touches every part of the system that receives and processes network packets... which is a good chunk of the system.  Hopefully it has zero meaningful effect.  A failure would be indicated by a system suddenly not being able to hear or react to a networking signal.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-e714de8_34e3d9dd.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-e714de8_34e3d9dd.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-e714de8.tar_34e3d9dd.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-e714de8.tar_34e3d9dd.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-e714de8_34e3d9dd.exe

**Comment by kasenvr** [comment]
Test case: Do things networking related and confirm nothing really changes. :)

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-d4cffb3.tar_34e3d9dd.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-d4cffb3.tar_34e3d9dd.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-d4cffb3_34e3d9dd.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-d4cffb3_34e3d9dd.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR613/Vircadia-Alpha-PR613-d4cffb3_34e3d9dd.exe

**Comment by kasenvr** [comment]
Did entity updates, tried talking to myself from multiple clients, more or less everything seems fine.

**Comment by kasenvr** [comment]
Also, there are missing copyright headers for the top of every file. Just need to append under the High Fidelity copyrights:

`Copyright 2020 Vircadia contributors.`

**Review by kasenvr — APPROVED** [review]
Upgrading my prior testing approval to full approval after hearing that this is a pretty safe PR. :)

**Pull request merged**